crates/matrix-sdk-crypto/src/store/memorystore.rs

```diff
@@ -272,47 +305,75 @@ impl CryptoStore for MemoryStore {
 
     async fn inbound_group_session_counts(
         &self,
-        _backup_version: Option<&str>,
+        backup_version: Option<&str>,
     ) -> Result<RoomKeyCounts> {
-        let backed_up =
-            self.get_inbound_group_sessions().await?.into_iter().filter(|s| s.backed_up()).count();
+        let backed_up = if let Some(backup_version) = backup_version {
+            self.get_inbound_group_sessions_and_backed_up_to()
+                .await?
+                .into_iter()
+                .filter(|(_, o)| *o == backup_version)
+                .count()
+        } else {
+            0
+        };
 
         Ok(RoomKeyCounts { total: self.inbound_group_sessions.count(), backed_up })
     }
 
     async fn inbound_group_sessions_for_backup(
         &self,
-        _backup_version: &str,
+        backup_version: &str,
         limit: usize,
     ) -> Result<Vec<InboundGroupSession>> {
         Ok(self
-            .get_inbound_group_sessions()
+            .get_inbound_group_sessions_and_backed_up_to()
             .await?
             .into_iter()
-            .filter(|s| !s.backed_up())
+            .filter_map(
+                |(session, backed_up_to)| {
+                    if backed_up_to != backup_version {
+                        Some(session)
+                    } else {
+                        None
+                    }
+                },
+            )
             .take(limit)
             .collect())
     }
 
     async fn mark_inbound_group_sessions_as_backed_up(
         &self,
-        _backup_version: &str,
+        backup_version: &str,
         room_and_session_ids: &[(&RoomId, &str)],
     ) -> Result<()> {
+        let mut inbound_group_sessions_backed_up_to =
+            self.inbound_group_sessions_backed_up_to.write().unwrap();
+
         for (room_id, session_id) in room_and_session_ids {
             let session = self.inbound_group_sessions.get(room_id, session_id);
             if let Some(session) = session {
                 session.mark_as_backed_up();
+
+                inbound_group_sessions_backed_up_to
+                    .entry((*room_id).to_owned())
+                    .or_default()
+                    .entry((*session_id).to_owned())
```

Comment:
Shirley the second `entry` could be replaced by an `.insert(session_id.to_owned(), backup_version.to_owned())`.

Reply:
Fixed in c8b1c59cc5 , and don't call me Shirley.